crates/e2e/macro/src/codegen.rs

```diff
@@ -98,28 +99,15 @@ impl InkE2ETest {
             };
 
         let mut already_built_contracts = already_built_contracts();
-        if already_built_contracts.is_empty() {
-            // Build all of them for the first time and initialize everything
-            BUILD_ONCE.call_once(|| {
-                tracing_subscriber::fmt::init();
-                for manifest_path in contracts_to_build_and_import {
-                    let dest_wasm = build_contract(&manifest_path);
-                    let _ = already_built_contracts.insert(manifest_path, dest_wasm);
-                }
-                set_already_built_contracts(already_built_contracts.clone());
-            });
-        } else if !already_built_contracts.is_empty() {
-            // Some contracts have already been built and we check if the
-            // `additional_contracts` for this particular test contain ones
-            // that haven't been build before
-            for manifest_path in contracts_to_build_and_import {
-                if already_built_contracts.get(&manifest_path).is_none() {
-                    let dest_wasm = build_contract(&manifest_path);
-                    let _ = already_built_contracts.insert(manifest_path, dest_wasm);
-                }
-            }
-            set_already_built_contracts(already_built_contracts.clone());
+        // Some contracts have already been built and we check if the
+        // `additional_contracts` for this particular test contain ones
+        // that haven't been build before
+        for manifest_path in contracts_to_build_and_import {
+            already_built_contracts
+                .entry(manifest_path.clone())
+                .or_insert_with(|| build_contract(&manifest_path));
```

Comment:
For tests running all in parallel, this might now result in all tests initially attempting to build the root contract simultaneously. I assume they would all just block on the lockfile while the first one builds, which is okay but negates the point of having the `already_built_contracts`. This mechanism itself might now be unnecessary since the build now is very fast when there have been no changes so we might even be able to remove this entirely.